File: batches_csharp/batch_05_function_extraction/compare.cs
using System;

class Palindrome {
    static string Normalize(string s) => s.ToLower().Replace(" ", "");
    static bool CharsMatch(string s, int l, int r) => s[l] == s[r];

    static bool IsPalindrome(string s) {
        s = Normalize(s);
        int left = 0, right = s.Length - 1;
        while (left < right) {
            if (!CharsMatch(s, left, right)) return false;
            left++; right--;
        }
        return true;
    }

    static void Main() {
        Console.WriteLine(IsPalindrome("A man a plan a canal Panama"));
        Console.WriteLine(IsPalindrome("hello"));
    }
}

File: batches_java/batch_05_function_extraction/compare.java
public class Palindrome {
    private static String normalize(String s) {
        return s.toLowerCase().replace(" ", "");
    }

    private static boolean charsMatch(String s, int left, int right) {
        return s.charAt(left) == s.charAt(right);
    }

    public static boolean isPalindrome(String s) {
        s = normalize(s);
        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (!charsMatch(s, left, right)) return false;
            left++;
            right--;
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(isPalindrome("A man a plan a canal Panama"));
        System.out.println(isPalindrome("hello"));
    }
}
